runtime/src/main/cpp/CyclicCollector.cpp

```diff
@@ -0,0 +1,450 @@
+           sideRefCounts_[root] = 0;
+         }
+         while (toVisit.size() > 0)  {
+           if (atomicGet(&mutatedAtomics_) != 0) {
+             COLLECTOR_LOG("restarted during rootset visit\n")
+             restartCount++;
+             goto restart;
+           }
+           auto* obj = toVisit.front();
+           toVisit.pop_front();
+           COLLECTOR_LOG("visit %s%p\n", isAtomicReference(obj) ? "atomic " : "", obj);
+           auto* objContainer = obj->container();
+           if (objContainer == nullptr) continue;  // Permanent object.
+           RuntimeCheck(objContainer->shareable(), "Must be shareable");
+           if (visited.count(obj) == 0) {
+             visited.insert(obj);
+             traverseObjectFields(obj, [&toVisit, obj, this](ObjHeader** location) {
+                ObjHeader* ref = *location;
+                if (ref != nullptr) {
+                  COLLECTOR_LOG("object field %p in %p\n", ref, obj);
+                  // We shall not account for edges inside the same frozen container, unless it originates
+                  // from an atomic reference.
+                  if (isAtomicReference(obj) || (obj->container() != ref->container())) {
+                    COLLECTOR_LOG("counting %p -> %p\n", obj, ref)
+                    sideRefCounts_[ref]++;
+                  } else {
+                    COLLECTOR_LOG("not counting %p -> %p\n", obj, ref)
+                  }
+                  toVisit.push_back(ref);
+                }
+             });
+           }
+         }
+         // Now find all elements with external references, and mark objects reachable from them as non suitable
+         // for collection by setting their side reference count to -1.
+         toVisit.clear();
+         for (auto it: sideRefCounts_) {
+           auto* obj = it.first;
+           auto* objContainer = obj->container();
+           if (objContainer == nullptr) continue;  // Permanent object.
+           RuntimeCheck(it.second <= objContainer->refCount(), "Must properly count inner refs");
+           if (it.second != objContainer->refCount()) {
+             COLLECTOR_LOG("for %p mismatched RC: %d vs %d, adding as possible root\n", obj, it.second, objContainer->refCount())
+             toVisit.push_back(it.first);
+           }
+         }
+         visited.clear();
+         while (toVisit.size() > 0)  {
+           auto* obj = toVisit.front();
+           toVisit.pop_front();
+           auto* objContainer = obj->container();
+           if (objContainer == nullptr) continue;  // Permanent object.
+           RuntimeCheck(objContainer->shareable(), "Must be shareable");
+           sideRefCounts_[obj] = -1;
+           visited.insert(obj);
+           if (atomicGet(&mutatedAtomics_) != 0) {
+             COLLECTOR_LOG("restarted during reachable visit\n")
+             restartCount++;
+             goto restart;
+           }
+           traverseObjectFields(obj, [&toVisit, &visited](ObjHeader** location) {
```

Comment:
One more data race (reproducible with sleeps as in https://github.com/JetBrains/kotlin-native/commit/e0d016b13287ad1c02d12c3e575de823475f5fea):
```kotlin
import kotlin.native.concurrent.*
  
class Node(val value: Any?)

fun createRoot(): AtomicReference<Any?> {
    val ref1 = AtomicReference<Any?>(null)
    val ref2 = AtomicReference<Any?>(null)

    ref1.value = Node(ref2).freeze()
    ref2.value = Any().freeze()

    return ref1
}

fun main() {
    val ref1 = createRoot()
    kotlin.native.internal.GC.collect()
    kotlin.native.internal.GC.collectCyclic()

    Worker.current.park(100 * 1000L)
    val node = ref1.value as Node
    ref1.value = null
    Worker.current.park(500 * 1000L)

    Worker.start().executeAfter(0L, {
        kotlin.native.internal.GC.collect()
    }.freeze())
    Worker.current.park(500 * 1000L)

    val ref2 = node.value as AtomicReference<Any?>
    println("VALUE: ${ref2.value}")
}
```